File: Data Structures/Linked List/linkedlist.cpp
#include <bits/stdc++.h>
using namespace std;
class Node
{
public:
    int data;
    Node *next;
};
Node *getNewNode(int data)
{
    Node *newNode = new Node();
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}
Node *head;
void printList()
{
    if (!head)
    {
        cout << "Head pointer is NULL" << endl;
        return;
    }
    Node *newNode = head;
    while (newNode)
    {
        cout << newNode->data << " ";
        newNode = newNode->next;
    }
    cout << endl;
}
void insertAt1(int data)
{
    Node *newNode = getNewNode(data);
    if (!head)
    {
        head = newNode;
        printList();
        return;
    }
    newNode->next = head;
    head = newNode;
    printList();
}
void insertAtLast(int data)
{
    Node *newNode = head;
    while (newNode->next)
    {
        newNode = newNode->next;
    }
    Node *newNode1 = getNewNode(data);
    newNode->next = newNode1;
    printList();
}
void insertAtN(int data, int position)
{
    if (position == 1)
    {
        insertAt1(data);
        return;
    }
    Node *newNode = head;
    for (int i = 0; i < position - 2; i++)
    {
        newNode = newNode->next;
    }
    Node *newNode1 = getNewNode(data);
    newNode1->next = newNode->next;
    newNode->next = newNode1;
    printList();
}
void deleteAt1()
{
    if (head)
    {
        Node *newNode = head;
        head = newNode->next;
        delete newNode;
        printList();
    }
}
void deleteAtLast()
{
    Node *newNode = head;
    while (newNode->next->next)
    {
        newNode = newNode->next;
    }
    Node *newNode1 = newNode->next;
    newNode->next = NULL;
    delete newNode1;
    printList();
}
void deleteAtN(int position)
{
    if (position == 1)
    {
        deleteAt1();
        return;
    }
    Node *newNode = head;
    for (int i = 0; i < position - 2; i++)
    {
        newNode = newNode->next;
    }
    Node *newNode1 = newNode->next;
    newNode->next = newNode1->next;
    delete newNode1;
    printList();
}
void recursive_reverse_print(Node *head_ref)
{
    if (!head_ref)
    {
        return;
    }
    recursive_reverse_print(head_ref->next);
    cout << head_ref->data << " ";
}
void reverse()
{
    Node *current = head;
    Node *prev = NULL;
    Node *tail;
    while (current)
    {
        tail = current->next;
        current->next = prev;
        prev = current;
        current = tail;
    }
    head = prev;
    printList();
}
int main()
{
    head = NULL;
    insertAt1(10);
    insertAtLast(20);
    insertAt1(30);
    insertAtN(40, 2);
    insertAtN(50, 3);
    insertAtN(60, 4);
    cout << "Reversed: ";
    recursive_reverse_print(head);
    cout << endl;
    reverse();
    deleteAt1();
    deleteAt1();
    deleteAtLast();
    deleteAtN(2);
    deleteAtN(1);
    deleteAt1();
}

File: Udemy Course/Linked List/linkedlist.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

    def __repr__(self):
        return f"Node data: {self.data}"


class LinkedList:
    def __init__(self):
        self.head = None

    def __repr__(self):
        nodes = []
        current = self.head
        while current:
            if current is self.head:
                nodes.append(f"[Head: {current.data}]")
            elif current.next is None:
                nodes.append(f"[Tail: {current.data}]")
            else:
                nodes.append(f"[{current.data}]")
            current = current.next
        return f"{'-> '.join(nodes)} Length: {len(self)}"

    def __len__(self):
        new_node = self.head
        count = 1
        while new_node.next:
            count += 1
            new_node = new_node.next
        return count

    def push(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node
        print(self)

    def append(self, data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
            return

        new_node.next = None
        new_node1 = self.head
        while new_node1.next:
            new_node1 = new_node1.next
        new_node1.next = new_node
        print(self)

    def insert(self, position, data):
        new_node = Node(data)
        if position == 0:
            self.push(data)
            return

        new_node1 = self.head
        for i in range(position - 2):
            new_node1 = new_node1.next
        new_node.next = new_node1.next
        new_node1.next = new_node
        print(self)

    def pop(self):
        new_node = self.head
        for i in range(len(self) - 2):
            new_node = new_node.next
        new_node1 = new_node.next
        new_node.next = new_node1.next
        del new_node1
        print(self)

    def removeAt0(self):
        new_node = self.head
        self.head = new_node.next
        del new_node
        print(self)

    def removeAtn(self, position):
        new_node = self.head
        if position == 0:
            self.removeAt0()
            return
        for i in range(position - 2):
            new_node = new_node.next
        new_node1 = new_node.next
        new_node.next = new_node1.next
        del new_node1
        print(self)

    def split_list(self, l_list):
        mid = len(l_list) // 2
        initial_head = self.head
        new_node = self.head
        left = LinkedList()
        right = LinkedList()
        for i in range(len(l_list)):
            if i < mid:
                left.append(new_node.data)
            else:
                right.append(new_node.data)
            new_node = new_node.next
        print("LEFT: ", left)
        print("RIGHT: ", right)
        self.head = initial_head
        return left, right

    def reverse(self):
        # Actually reverses the list instead of just printing from tail end
        curr = self.head
        prev = None
        while curr:
            tail = curr.next
            curr.next = prev
            prev = curr
            curr = tail
        self.head = prev
        print(self)

    @staticmethod
    def recursive_reverse_print(head):
        # Doesn't actually reverse the list, Just prints it from tail end
        if not head:
            return
        LinkedList.recursive_reverse_print(head.next)
        print(head.data, end=" ")


if __name__ == "__main__":
    obj = LinkedList()
    obj.push(15)
    obj.push(20)
    obj.push(76)
    obj.push(12)
    obj.append(75)
    obj.append(98)
    obj.insert(0, 999)
    obj.insert(1, 7878)
    obj.insert(8, 1714)
    # obj.split_list(obj)
    # print(obj)
    obj.reverse()
    print("Reversed: ", end="")
    LinkedList.recursive_reverse_print(obj.head)
    print('\n')
    obj.pop()
    obj.removeAtn(1)
    obj.removeAt0()
    obj.removeAt0()
